cpp/src/parquet/arrow/reader_internal.cc

```diff
@@ -55,17 +59,520 @@ using arrow::TimestampArray;
+    *out = ::arrow::null();
+    return Status::OK();
+  }
+
+  switch (primitive.physical_type()) {
+    case ParquetType::BOOLEAN:
+      *out = ::arrow::boolean();
+      break;
+    case ParquetType::INT32:
+      RETURN_NOT_OK(FromInt32(*logical_type, out));
+      break;
+    case ParquetType::INT64:
+      RETURN_NOT_OK(FromInt64(*logical_type, out));
+      break;
+    case ParquetType::INT96:
+      *out = ::arrow::timestamp(::arrow::TimeUnit::NANO);
+      break;
+    case ParquetType::FLOAT:
+      *out = ::arrow::float32();
+      break;
+    case ParquetType::DOUBLE:
+      *out = ::arrow::float64();
+      break;
+    case ParquetType::BYTE_ARRAY:
+      RETURN_NOT_OK(FromByteArray(*logical_type, out));
+      break;
+    case ParquetType::FIXED_LEN_BYTE_ARRAY:
+      RETURN_NOT_OK(FromFLBA(*logical_type, primitive.type_length(), out));
+      break;
+    default: {
+      // PARQUET-1565: This can occur if the file is corrupt
+      return Status::IOError("Invalid physical column type: ",
+                             TypeToString(primitive.physical_type()));
+    }
+  }
+  return Status::OK();
+}
+
+struct SchemaTreeContext {
+  SchemaManifest* manifest;
+  ArrowReaderProperties properties;
+  const SchemaDescriptor* schema;
+
+  void LinkParent(const SchemaField* child, const SchemaField* parent) {
+    manifest->child_to_parent[child] = parent;
+  }
+
+  void RecordLeaf(const SchemaField* leaf) {
+    manifest->column_index_to_field[leaf->column_index] = leaf;
+  }
+};
+
+Status GetTypeForNode(int column_index, const schema::PrimitiveNode& primitive_node,
+                      SchemaTreeContext* ctx, std::shared_ptr<DataType>* out) {
+  std::shared_ptr<DataType> storage_type;
+  RETURN_NOT_OK(GetPrimitiveType(primitive_node, &storage_type));
+  if (ctx->properties.read_dictionary(column_index)) {
+    *out = ::arrow::dictionary(::arrow::int32(), storage_type);
+  } else {
+    *out = storage_type;
+  }
+  return Status::OK();
+}
+
+Status NodeToSchemaField(const Node& node, int16_t max_def_level, int16_t max_rep_level,
+                         SchemaTreeContext* ctx, const SchemaField* parent,
+                         SchemaField* out);
+
+Status GroupToSchemaField(const GroupNode& node, int16_t max_def_level,
+                          int16_t max_rep_level, SchemaTreeContext* ctx,
+                          const SchemaField* parent, SchemaField* out);
+
+Status PopulateLeaf(int column_index, const std::shared_ptr<Field>& field,
+                    int16_t max_def_level, int16_t max_rep_level, SchemaTreeContext* ctx,
+                    const SchemaField* parent, SchemaField* out) {
+  out->field = field;
+  out->column_index = column_index;
+  out->max_definition_level = max_def_level;
+  out->max_repetition_level = max_rep_level;
+  ctx->RecordLeaf(out);
+  ctx->LinkParent(out, parent);
+  return Status::OK();
+}
+
+inline bool EndswithTuple(const std::string& str) {
```

Comment:
I think this code was moved from elsewhere but you could replace this with `boost::algorithm::ends_with`

Reply:
Cool, doing that then =)